sql/handler.cc

```diff
@@ -1734,17 +1734,38 @@ int ha_commit_trans(THD *thd, bool all)
     */
     if (run_wsrep_hooks)
     {
-      // This commit involves more than one storage engine and requires
-      // two phases, but some engines don't support it.
-      // Issue a message to the client and roll back the transaction.
+      // This commit involves storage engines that do not support two phases.
+      // We allow read only transactions to such storage engines but not
+      // read write transactions.
       if (trans->no_2pc && rw_ha_count > 1)
       {
-	// REPLACE|INSERT INTO ... SELECT uses TOI for MyISAM|Aria
-	if (WSREP(thd) && thd->wsrep_cs().mode() != wsrep::client_state::m_toi)
-	{
-          my_message(ER_ERROR_DURING_COMMIT, "Transactional commit not supported "
-                     "by involved engine(s)", MYF(0));
-          error= 1;
+        uint no2pc_rw_ha_count=0;
```

Comment:
Or maybe 
```
bool wsrep_have_no2pc_rw_ha(Ha_trx_info*)
```
to avoid counting all handlertons and write
```
if (trans->no_2pc && rw_ha_count > 1 && wsrep_have_no2pc_rw_ha(ha_info))
{
...
```

Reply:
And `uint no2pc_rw_ha_count=0;` is a bit misleading as we always break on the first found. We need something like: `bool has_no2pc_rw_ha;` as a flag, not a counter.

Reply:
Good points, I will use additional function and just boolean.